processor/src/lib.rs

```diff
@@ -227,33 +216,83 @@ impl Process {
-        for &op in block.op_batches()[0].ops() {
-            self.execute_op(op)?;
-            self.decoder.execute_user_op(op);
-        }
+        self.execute_op_batch(&block.op_batches()[0])?;
 
         // if the span contains more operation batches, execute them. each additional batch is
         // preceded by a RESPAN operation; executing RESPAN operation does not change the state
         // of the stack
         for op_batch in block.op_batches().iter().skip(1) {
             self.decoder.respan(op_batch);
             self.execute_op(Operation::Noop)?;
-            for &op in op_batch.ops() {
+            self.execute_op_batch(op_batch)?;
+        }
+
+        self.end_span_block(block)
+    }
+
+    /// Executes all operations in an [OpBatch]. This also ensure that all alignment rules are
+    /// satisfied by executing NOOPs as needed. Specifically:
+    /// - If an operation group ends with an operation carrying an immediate value, a NOOP is
+    ///   executed after it.
+    /// - If the number of groups in a batch is not a power of 2, NOOPs are executed (one per
+    ///   group) to bring it up to the next power of two (e.g., 3 -> 4, 5 -> 8).
+    #[inline(always)]
+    fn execute_op_batch(&mut self, batch: &OpBatch) -> Result<(), ExecutionError> {
+        let op_counts = batch.op_counts();
+        let mut op_idx = 0;
+        let mut group_idx = 0;
+        let mut next_group_idx = 1;
+
+        // execute operations in the batch one by one
+        for &op in batch.ops() {
+            if op.is_decorator() {
+                // if the operation is a decorator, it has no side effects and, thus, we don't
+                // need to decode it
                 self.execute_op(op)?;
-                self.decoder.execute_user_op(op);
+                continue;
+            }
+
+            // decode and execute the operation
+            self.decoder.execute_user_op(op);
+            self.execute_op(op)?;
+
+            // if the operation carries an immediate value, the value is stored at the next group
+            // pointer; so, we advance the pointer to the following group
+            let has_imm = op.imm_value().is_some();
+            if has_imm {
+                next_group_idx += 1;
+            }
+
+            // determine if we've executed all non-decorator operations in a group
+            if op_idx == op_counts[group_idx] - 1 {
+                // if so, move to the next group and reset operation index
+                group_idx = next_group_idx;
+                next_group_idx += 1;
+                op_idx = 0;
+
+                // if the operation carries immediate value, it cannot be the last operation in
+                // the group - so, just execute a NOOP after it
+                if has_imm {
+                    // an operation with immediate value cannot be in the 9th position
+                    debug_assert!(op_idx < OP_GROUP_SIZE - 1, "invalid op index");
```

Comment:
This will always be true, since `op_idx` is set to zero above. Maybe this block was intended to be before the index updates above?

Reply:
Ah yes! I had them swapped at some point, but then arranged them as they are now without revisiting the logic. Good catch!

Reply:
Fixed!